extensions/amp-form/0.1/amp-form.js

```diff
@@ -909,24 +924,32 @@ export class AmpForm {
    */
   handleXhrSubmitSuccess_(response) {
     const json = /** @type {!Promise<!JsonObject>} */ (response.json());
-    return this.handleSubmitSuccess_(json).then(() => {
+    return this.handleSubmitSuccess_(
+      json,
+      /** getSsrEventResponse */ false
+    ).then(() => {
       this.triggerFormSubmitInAnalytics_('amp-form-submit-success');
       this.maybeHandleRedirect_(response);
     });
   }
 
   /**
    * Transition the form to the submit success state.
+   * If getSsrEventResponse is true, exposes the response.body in
+   * the SUBMIT_SUCCESS event; otherwise exposes the resolved jsonPromise.
    * @param {!Promise<!JsonObject>} jsonPromise
+   * @param {boolean} getSsrEventResponse
    * @return {!Promise}
    * @private visible for testing
    */
-  handleSubmitSuccess_(jsonPromise) {
+  handleSubmitSuccess_(jsonPromise, getSsrEventResponse) {
     return jsonPromise.then(
       json => {
+        // HTTP response payload returned by the amp-form endpoint from SSR
+        const eventResponse = getSsrEventResponse ? json['body'] : json;
```

Comment:
We might be able to make this cleaner by doing this earlier e.g. in `handleSsrTemplateResponse_` instead of plumbing a boolean down several method calls.

We should also change this first parameter from `Promise<!JsonObject>` to just `!JsonObject`.

```js
// SSR case.
return this.handleSubmitSuccess_(
  response, // Optional: Change this to response['html'] and add `|string` to affected types.
  response['body']
);

// CSR case.
return response.json().then(json => this.handleSubmitSuccess_(json, json));

// New signature.
handleSubmitSuccess_(result, eventData) {
  ...
  this.renderTemplate_(result || {}).then(() => {
    this.triggerAction_(FormEvents.SUBMIT_SUCCESS, eventData);
    ...
  });
}
```

Reply:
This might be simpler than [my suggestion](https://github.com/ampproject/amphtml/pull/25242#discussion_r338721178)

Reply:
This is so much nicer thanks to both of you! PTAL when you get a chance.